homeassistant/helpers/event.py

```diff
@@ -1126,23 +1125,40 @@ def async_track_point_in_utc_time(
     point_in_time: datetime,
 ) -> CALLBACK_TYPE:
     """Add a listener that fires once after a specific point in UTC time."""
-    # Ensure point_in_time is UTC
-    utc_point_in_time = dt_util.as_utc(point_in_time)
-
     # Since this is called once, we accept a HassJob so we can avoid
     # having to figure out how to call the action every time its called.
     job = action if isinstance(action, HassJob) else HassJob(action)
 
-    cancel_callback = hass.loop.call_at(
-        hass.loop.time() + point_in_time.timestamp() - time.time(),
-        hass.async_run_hass_job,
-        job,
-        utc_point_in_time,
-    )
+    cancel_callback: Optional[asyncio.TimerHandle] = None
+
+    @callback
+    def run_action() -> None:
+        """Call the action."""
+        nonlocal cancel_callback
+
+        now = track_point_in_utc_time_now()
+
+        # Depending on the available clock support (including timer hardware
+        # and the OS kernel) it can happen that we fire a little bit too early
+        # as measured by utcnow(). That is bad when callbacks have assumptions
+        # about the current time. Thus, we rearm the timer for the remaining
+        # time.
+        delta = (point_in_time - now).total_seconds()
+        if delta > 0:
+            _LOGGER.debug("Called %f seconds too early, rearming", delta)
+
+            cancel_callback = hass.loop.call_later(delta, run_action)
+            return
+
+        hass.async_run_hass_job(job, now)
```

Comment:
I think we should preserve the existing behavior of calling the job back with `utc_point_in_time`

```suggestion
        # Depending on the available clock support (including timer hardware
        # and the OS kernel) it can happen that we fire a little bit too early
        # as measured by utcnow(). That is bad when callbacks have assumptions
        # about the current time. Thus, we rearm the timer for the remaining
        # time.
        delta = point_in_time.timestamp() - time.time()
        if delta > 0:
            _LOGGER.debug("Called %f seconds too early, rearming", delta)

            cancel_callback = hass.loop.call_later(delta, run_action)
            return

        hass.async_run_hass_job(job, utc_point_in_time)
```

Reply:
I had a few small reasons for preferring to pass the current time rather than the scheduled time:

* It is slightly more accurate.
* It is new information while `point_in_time` is something the caller already knows.
* It ensures progress, for example if naively using `now + now%10` to schedule a timer every ten seconds.
* It seems more useful with hibernation.

All are minor so if you have some points for doing it the other way I could easily be swayed?

(BTW, we used to pass the current time before the event loop rework so it is not a groundbreaking change)

Reply:
I should have given some more backstory here, and why I think it would be a breaking change if we don't preserve the existing behavior:

When we worked on this before, we found that some callers would use the time that was returned with the event was used to schedule the next event and so on.

If we return the time the event fired, instead of the time that they expect it to fire, the next event is going to be scheduled slightly later, and the one after that slightly later then that until we finally schedule past the second and potentially miss an expected run.

When we previously refactored, we left it returning the expected time since the thought was that if they wanted `utcnow()`, they could call it themselves when the callback ran.

Reply:
Hmm, I must have missed that discussion. Situations like #39335 were what I was thinking about but now you say there are other callers making the opposite assumption. Potato, potahto ... I think it would be a bit silly to revert a breaking change so I will use your suggestion.

Reply:
So regardless if we should or shouldn't do, shouldn't that be another PR? It's an impactful change and it might be that more people have opinions.

Reply:
That was tricky.

I now reverted so we pass the current time in `async_track_utc_time_change()` and the requested time in `async_track_point_in_utc_time()`. This seems confusing but it is how it works today; see https://github.com/home-assistant/core/blob/3661d8380af5035d0c2d0cbce4b89a5de099b902/homeassistant/helpers/event.py#L1349-L1350 and https://github.com/home-assistant/core/blob/3661d8380af5035d0c2d0cbce4b89a5de099b902/homeassistant/helpers/event.py#L1140

While getting that in place I noticed that the time pattern tests for DST changes are currently wrong because they use a moving year but DST changes do not fall on the same date each year.

Getting time to stand still so these tests could be fixed required some additional patching of `utcnow()` and this means that `time.time()` can no longer be used (without even more patching). I believe that is okay due to the recent `utcnow()` optimizations?

I pushed two further commits with these changes.

Reply:
You are correct that `utcnow()` is now much faster, ideally we do use `time.time()` since its still almost 4x faster than `utcnow()`